src/picosvg/svg.py

```diff
@@ -944,6 +948,69 @@ def _apply_gradient_translation(self, inplace=False):
             else:
                 del el.attrib["gradientTransform"]
 
+    def _resolve_gradient_templates(self, inplace=False):
+        # Gradients can have an 'href' attribute that specifies another gradient as
+        # a template, inheriting its attributes and/or stops when not already defined:
+        # https://www.w3.org/TR/SVG/pservers.html#PaintServerTemplates
+        if not inplace:
+            svg = SVG(copy.deepcopy(self.svg_root))
+            svg._resolve_gradient_templates(inplace=True)
+            return svg
+
+        href_attr = _xlink_href_attr_name()
+        el_by_id = {el.attrib["id"]: el for el in self.xpath(".//svg:*[@id]")}
+
+        def resolve_gradient_template(gradient):
+            ref = gradient.attrib[href_attr]
+            if not ref.startswith("#"):
+                raise ValueError(f"Only use #fragment supported, reject {ref}")
+
+            template = el_by_id.get(ref[1:])
+            if template is None:
+                raise ValueError(f"No element has id '{ref[1:]}'")
+
+            template_tag = strip_ns(template.tag)
+            if template_tag not in _GRADIENT_CLASSES:
+                raise ValueError(
+                    f"Referenced element with id='{ref[1:]}' has unexpected tag: "
+                    f"expected linear or radialGradient, found '{template_tag}'"
+                )
+
+            # recurse if template references another template
+            if template.attrib.get(href_attr):
+                resolve_gradient_template(template)
+
+            for attr_name in _GRADIENT_ATTRS[strip_ns(gradient.tag)]:
+                if attr_name in template.attrib and attr_name not in gradient.attrib:
+                    gradient.attrib[attr_name] = template.attrib[attr_name]
+
+            if len(gradient) == 0:
```

Comment:
len(gradient) threw me; maybe a comment? - `# only copy stops if we don't have our own` or some such